File: haskell/ac15.hs
import Data.Array
import Data.Char
import qualified Data.Map as Map
import qualified Data.Set as Set

cut p list =
  let (hd, (_:tl)) = break p list
  in (hd, tl)

split p list =
  let (hd, tl) = break p list
  in case tl of
    [] -> [hd]
    _:tl -> hd:split p tl

addCharToHash current_value char =
  (current_value + ord char) * 17 `mod` 256

hash = foldl addCharToHash 0

stripRemove [] _set accu = accu

stripRemove (hd:tl) set accu =
  if last hd == '-' then
    stripRemove tl (Set.insert (init hd) set) accu
  else
    let (label, focus_str) = cut (== '=') hd
        accu' =
          if Set.member label set then
            accu
          else
            (label, read focus_str):accu
    in stripRemove tl set accu'

insertLen :: (Array Int [String], Map.Map String Int) -> (String, Int) -> (Array Int [String], Map.Map String Int)
insertLen (boxes, focuses) (label, focus) =
  let boxes' =
        if Map.member label focuses then
          boxes
        else
          accum (flip (:)) boxes [(hash label, label)]
  in (boxes', Map.insert label focus focuses)

measureLen focuses (i, label) =
  i * focuses Map.! label

sumBox focuses (i, lens) =
  (i + 1) * sum (map (measureLen focuses) (zip [1..] (reverse lens)))

main = do
  contents <- getContents
  let instructions_str = split (== ',') (filter (/= '\n') contents)
  let result_part1 = sum (map hash instructions_str)
  putStrLn("Part 1: " ++ show result_part1)
  let set_instructions = stripRemove (reverse instructions_str) Set.empty []
  let init_boxes = array (0, 255) [(i, []) | i <- [0 .. 255]]
  let (boxes, focuses) =
        foldl insertLen (init_boxes, Map.empty) set_instructions
  let result_part2 = sum (map (sumBox focuses) (assocs boxes))
  putStrLn("Part 2: " ++ show result_part2)

File: ocaml/AC15.ml
let add_char_to_hash current_value char =
  (current_value + int_of_char char) * 17 mod 256

let hash s =
  Seq.fold_left add_char_to_hash 0 (String.to_seq s)

module StringSet = Set.Make (String)

module StringMap = Map.Make (String)

let strip_remove (removed, accu) step =
  if String.ends_with ~suffix:"-" step then
    let label = String.sub step 0 (String.length step - 1) in
    StringSet.add label removed, accu
  else
    let (label, focus) =
      Scanf.sscanf step "%s@=%d" (fun label focus -> (label, focus)) in
    let accu =
      if StringSet.mem label removed then
        accu
      else
        (label, focus) :: accu in
    removed, accu

let execute_set_step boxes focuses (label, focus) =
  if not (StringMap.mem label focuses) then
    begin
      let index = hash label in
      boxes.(index) <- label :: boxes.(index)
    end;
  StringMap.add label focus focuses

let add_len_focusing_power focuses (i, total) label =
  i + 1, total + i * StringMap.find label focuses

let add_box_focusing_power focuses (i, total) lens =
  let _, box_total =
    List.fold_left (add_len_focusing_power focuses) (1, 0) (List.rev lens) in
  i + 1, total + i * box_total

let () =
  let input = Common.input_lines () in
  let s = String.concat "" (List.of_seq input) in
  let steps = String.split_on_char ',' s in
  let hashes = List.map hash steps in
  let result_part1 = List.fold_left ( + ) 0 hashes in
  Format.eprintf "Part 1: %d@." result_part1;
  let _, set_steps =
    List.fold_left strip_remove (StringSet.empty, []) (List.rev steps) in
  let boxes = Array.make 256 [] in
  let focuses =
    List.fold_left (execute_set_step boxes) StringMap.empty set_steps in
  let _, result_part2 =
    Array.fold_left (add_box_focusing_power focuses) (1, 0) boxes in
  Format.eprintf "Part 2: %d@." result_part2
